feat: add revenue sharing specs (#47)
defi-wonderland/specs: specs/protocol/jovian/predeploys.md

```solidity
event MinWithdrawalAmountUpdated(uint256 oldWithdrawalAmount, uint256 newWithdrawalAmount)
```

#### `RecipientUpdated`

Emitted when the fee recipient for this vault is updated.

```solidity
event RecipientUpdated(address oldRecipient, address newRecipient)
```

#### `WithdrawalNetworkUpdated`

Emitted when the withdrawal network for this vault is updated.

```solidity
event WithdrawalNetworkUpdated(WithdrawalNetwork oldWithdrawalNetwork, WithdrawalNetwork newWithdrawalNetwork)
```

### Invariants

- Only `ProxyAdmin.owner()` is allowed to call the setter functions.
- If using the `FeeSplitter`, the withdrawal network MUST be set to `WithdrawalNetwork.L2` and the recipient
  MUST be set to the `FeeSplitter` predeploy address.
- The balance of the vault MUST be preserved between implementation upgrades.

## FeeSplitter

This contract splits the ETH it receives and sends the correct amounts to two designated addresses. It integrates with
the fee vault system by configuring each Fee Vault to use `WithdrawalNetwork.L2` and setting this predeploy as the
recipient in EVERY fee vault.

The contract manages two recipients:

- Recipient A

> - Recipient B

Reviewer: We need to rename these names everywhere. We should define something more specific to the usecase, for example: Revenue share recipient and revenue rest recipient

Or something similar

Reviewer 2: [59d00af](https://github.com/defi-wonderland/specs/pull/47/commits/59d00afeed9029c707bb06d978b72b64e07ff0e4)

Based on the configured share percentage for Recipient A, the contract sends the corresponding portion of fees
to Recipient A and sends the remaining amount to Recipient B.

### Functions

#### `disburseFees`

Initiates the routing flow by withdrawing the fees that each of the fee vaults has collected and sends the shares
to the appropriate addresses according to the configured percentage.

When attempting to withdraw from the vaults, it will check that they all have a balance equal to or larger than
their minimum withdrawal amount, that the withdrawal network is set to `WithdrawalNetwork.L2`, and that the recipient
of the vault is the `FeeSplitter`. The function MUST revert if any of these conditions is not met.
The function MUST emit the `NoFeesCollected` event if the contract doesn't have any funds after the vaults have been withdrawn,
which can happen in scenarios where the vaults have a minimum withdrawal amount of 0.

> On chains where any fee vault is not to be used and is not expected to maintain a balance in the contract,

Reviewer: I think this sentence is not clear enough

Reviewer 2: I think it can be removed, since if there is no fees it won't revert but it just won't withdraw anything.

it is necessary to set the unused fee vault's minimum withdrawal amount to 0 to prevent `disburseFees` from
reverting when it shouldn't.

Currently this is the case for `OperatorFeeVault` on some chains, so setting its minimum withdrawal amount to

> 0 is needed on those chains.

Reviewer 2: Not needed

> 0 is needed on those chains.

Reviewer 2: [59d00af](https://github.com/defi-wonderland/specs/pull/47/commits/59d00afeed9029c707bb06d978b72b64e07ff0e4)

```solidity
function disburseFees() external
```

- MUST emit `FeesDisbursed` event upon successful execution.
- MUST emit `NoFeesCollected` event if there are no funds available in the contract after the vaults have been withdrawn.

> - MUST revert if any vault has a balance below its minimum withdrawal amount.

Reviewer: This is wrong, we should not revert if there is some other vault with enough balance

Reviewer 2: [59d00af](https://github.com/defi-wonderland/specs/pull/47/commits/59d00afeed9029c707bb06d978b72b64e07ff0e4)

These contracts inherit the changes made to the `FeeVault` contract, meaning that they have storage
variables and setters instead of constants for the configuration values, and they are initializable.

Their configuration includes the withdrawal network and the recipient to which the fees are sent:

- **WithdrawalNetwork.L1**: Funds are withdrawn to an L1 address (default behavior)
- **WithdrawalNetwork.L2**: Funds are withdrawn to an L2 address

For those chains that choose to use the `FeeSplitter` predeploy, `WithdrawalNetwork.L2` as the withdrawal
network, and the `FeeSplitter` as the recipient MUST be set using the setter functions.

## FeeSplitter

This contract splits the funds it receives from the vaults using a configured `ISharesCalculator` compatible revenue shares calculator to determine which addresses should receive funds and in what amounts by querying `ISharesCalculator.getRecipientsAndAmounts`:

```solidity
function getRecipientsAndAmounts(
        uint256 _sequencerFeeVaultBalance,
        uint256 _baseFeeVaultBalance,
        uint256 _operatorFeeVaultBalance,
        uint256 _l1FeeVaultBalance)
        external
        view

>         returns (ShareInfo[] memory shareInfo);

Reviewer 2: Let's define the struct as well

Author: Addressed [here](https://github.com/defi-wonderland/specs/pull/47/commits/eafd5b0657301a085183d8cb3707e331785a4443).

- MUST set `feeDisbursementInterval` to `_feeDisbursementInterval`.
- MUST emit an `Initialized` event with the provided parameters.

#### `disburseFees`

Initiates the routing flow by withdrawing the fees that each of the fee vaults has collected and sends the shares
to the appropriate addresses according to the amounts returned by the set shares calculator.

When attempting to withdraw from the vaults, it checks that the withdrawal network is set to `WithdrawalNetwork.L2`,
and that the recipient of the vault is the `FeeSplitter`. It MUST revert if any of these conditions is not met.
It MUST only withdraw if the vault balance is greater than or equal to its minimum withdrawal amount.
In addition, it follows a `nonReentrant` pattern using `TSORE`d flags, to avoid receiving balance back
once the fees are being disbursed.

```solidity
function disburseFees() external
```

- MUST revert if not enough time has passed since the last successful execution.
- MUST revert if any vault has a recipient different from this contract.
- MUST revert if any vault has a withdrawal network different from `WithdrawalNetwork.L2`.
- MUST revert if total fees collected are 0.

>   withdrawal amount set.

Reviewer 2: This sentence here seems off

    - [`receive`](#receive)
    - [`setRevenueShareRecipient`](#setrevenuesharerecipient)
    - [`setNetFeeShareBP`](#setnetfeesharebp)
    - [`setGrossFeeShareBP`](#setgrossfeesharebp)
    - [`setRevenueRemainderRecipient`](#setrevenueremainderrecipient)
    - [`setFeeDisbursementInterval`](#setfeedisbursementinterval)
  - [Events](#events-1)
    - [`FeesDisbursed`](#feesdisbursed)
    - [`NoFeesCollected`](#nofeescollected)
    - [`FeesReceived`](#feesreceived)
    - [`NetFeeShareBPUpdated`](#netfeesharebpupdated)
    - [`GrossFeeShareBPUpdated`](#grossfeesharebpupdated)
    - [`Initialized`](#initialized)
    - [`RevenueShareRecipientUpdated`](#revenuesharerecipientupdated)
    - [`RevenueRemainderRecipientUpdated`](#revenueremainderrecipientupdated)
    - [`FeeDisbursementIntervalUpdated`](#feedisbursementintervalupdated)
- [Security Considerations](#security-considerations)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

| Name        | Address                                    | Introduced | Deprecated | Proxied |
| ----------- | ------------------------------------------ | ---------- | ---------- | ------- |

> | FeeSplitter | 0x420000000000000000000000000000000000001C | Jovian     | No         | Yes     |

Reviewer 3: Check with the PoC on what the correct address is

to the appropriate addresses according to the configured percentage.
The function MUST revert if the withdrawal is not set to `WithdrawalNetwork.L2`, or if the recipient set is not the `FeeSplitter`.
The function MUST withdraw only if the vault balance is greater than or equal to its minimum withdrawal amount.

When attempting to withdraw from the vaults, it will check that the withdrawal network is set to `WithdrawalNetwork.L2`, and that the recipient of the vault is the `FeeSplitter`. It MUST revert if any of these conditions are not met.
It MUST only withdraw if the vault balance is greater than or equal to its minimum withdrawal amount.
In addition, it will follow a `nonReentrant` pattern, to avoid receiving balance back once the fees are being disbursed.

```solidity
function disburseFees() external
```

- MUST revert if not enough time has passed since the last successful execution.
- MUST revert if any vault has a recipient different from this contract.
- MUST revert if any vault has a withdrawal network different from `WithdrawalNetwork.L2`.
- MUST withdraw the vault's fees balance if the vault's balance is equal to or greater than the minimum withdrawal amount set.

> - MUST set the `lastDisbursementTime` to the current block timestamp.

Reviewer 3: Maybe it is helpful to mention that this is updated only when `disburseFees` collects ETH

```solidity
event Initialized(
        address revenueShareRecipient,
        address revenueRemainderRecipient,
        uint40 feeDisbursementInterval,
        uint16 netFeeShareBP,
        uint16 grossFeeShareBP
    )
```

#### `RevenueShareRecipientUpdated`

Emitted when the revenue share recipient is successfully updated.

```solidity
event RevenueShareRecipientUpdated(address indexed oldRevenueShareRecipient, address indexed newRevenueShareRecipient)
```

#### `RevenueRemainderRecipientUpdated`

Emitted when the revenue remainder recipient is successfully updated.

```solidity
event RevenueRemainderRecipientUpdated(address indexed oldRevenueRemainderRecipient, address indexed newRevenueRemainderRecipient)
```

#### `FeeDisbursementIntervalUpdated`

Emitted when the minimum time interval between consecutive fee disbursements is successfully updated.

```solidity
event FeeDisbursementIntervalUpdated(uint256 oldFeeDisbursementInterval, uint256 newFeeDisbursementInterval)
```

## Security Considerations

- Given that vault recipients can now be updated, it's important to ensure that this can only be done by the appropriate address, namely `ProxyAdmin.owner()`.

> - Upgrading the vaults and making them compatible with the `FeeSplitter` incurs a process where you have to deploy the new implementations and properly configure the vaults, which introduces complexity and potential for errors. It is important to develop a solution, such as a contract to manage the entire upgrade process, simplifying the UX and reducing the risk of errors.

Reviewer 3: ```suggestion
- Upgrading the vaults and making them compatible with the `FeeSplitter` incurs a process that requires deploying the new implementations and properly configuring the vaults, which introduces complexity and potential for errors. It is important to develop a solution, such as a contract to manage the entire upgrade process, simplifying the UX and reducing the risk of errors.
```

- The balance of the vault MUST be preserved between implementation upgrades.

## FeeSplitter

This contract splits the ETH it receives and sends the correct amounts to two designated addresses. It integrates with
the fee vault system by configuring each Fee Vault to use `WithdrawalNetwork.L2` and setting this predeploy as the
recipient in EVERY fee vault.

The contract manages two recipients:

- Recipient A
- Recipient B

Based on the configured share percentage for Recipient A, the contract sends the corresponding portion of fees
to Recipient A and sends the remaining amount to Recipient B.

### Functions

#### `disburseFees`

Initiates the routing flow by withdrawing the fees that each of the fee vaults has collected and sends the shares
to the appropriate addresses according to the configured percentage.

```solidity
function disburseFees() external
```

- MUST emit `FeesDisbursed` event upon successful execution.
- MUST emit `NoFeesCollected` event if there are no funds available at the time of the call.
- MUST revert if not enough time has passed since the last successful execution.
- MUST send the appropriate amounts to the recipients.

> - The balance of the contract MUST be 0 after a successful execution.

Author: That's a pretty good point.

```

- MUST only be callable by `ProxyAdmin.owner()`
- MUST revert if `_shareBasisPoints` exceeds `BASIS_POINT_SCALE`.
- MUST emit a `RecipientAShareUpdated` event upon successful execution.

#### `setRecipientB`

Sets the address that should receive the remaining fees after the recipient A share has been deducted.

```solidity
function setRecipientB(address _recipientB) external
```

- MUST only be callable by `ProxyAdmin.owner()`
- MUST emit a `RecipientBUpdated` event upon successful execution.

#### `setFeeDisbursementInterval`

Sets the minimum time, in seconds, that must pass between consecutive calls to `disburseFees`.

```solidity
function setFeeDisbursementInterval(uint32 _newInterval) external
```

- MUST only be callable by `ProxyAdmin.owner()`
- MUST emit a `FeeDisbursementIntervalUpdated` event upon successful execution.

### Events

#### `FeesDisbursed`

Emitted when fees are successfully withdrawn from fee vaults and distributed to recipients.

```solidity
event FeesDisbursed(
        uint256 indexed disbursementTime, uint256 paidToRecipientA, uint256 paidToRecipientB, uint256 totalFeesDisbursed
    );
```

#### `NoFeesCollected`

> Emitted when `disburseFees` is called but there are no funds available in the fee vaults at the time of execution.

Author: Addressed [here](https://github.com/defi-wonderland/specs/pull/47/commits/66f095d92fd0de7a8c170dceae511a736607eda0).
